interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -472,39 +521,43 @@ func (pdb *PromptDB) HandleNewRule(metadata *prompting.Metadata, constraints *pr
 		if !(prompt.Snap == metadata.Snap && prompt.Interface == metadata.Interface) {
 			continue
 		}
-		matched, err := constraints.Match(prompt.Constraints.path)
+
+		matched, satisfied, denied, err := prompt.Constraints.applyRuleConstraints(constraints)
 		if err != nil {
+			// Should not occur, only error is if path pattern is malformed,
+			// which would have thrown an error while parsing, not now.
 			return nil, err
 		}
 		if !matched {
 			continue
 		}
-
-		// Record all allowed permissions at the time of match, in case a
-		// permission is denied and we need to send back a response.
-		allowedPermission := responseForInterfaceConstraintsOutcome(metadata.Interface, prompt.Constraints, outcome)
-
-		// See if the permission matches any of the prompt's remaining permissions
-		modified := prompt.Constraints.subtractPermissions(constraints.Permissions)
-		if !modified {
-			// No permission was matched
+		// Matched, so at least one permission was satisfied
```

Comment:
I'm not sure I understand the comment vs the code, matched doesn't meant that all outcomes are not deny? it seems the comment needs to be clarified/expanded

Reply:
Matched means the path pattern of the rule matched the path of the prompt, and at least one permission from the rule matched at least one permission from the prompt. Whether all or just some of the permissions were matched, and whether each was allowed or denied, is what the other return values of `matched, satisfied, denied, err := prompt.Constraints.applyRuleConstraints(constraints)` indicates.

But I agree this is rather confusing. `matched`, `satisfied`, and `denied` are not the same types, and there is some implication between each.

I think the complexity comes down to the interaction between the way `buildResponse`, `applyRuleConstraints`, and their callers interact. A lot of it is because of the old way these used to work, which is no longer the case, so they can be simplified. I'll work on that.